### 專案名稱: 第四屆機器學習百日馬拉松
### 功能描述: 第34天作業
### 版權所有: Dunk  
### 程式撰寫: Dunk  
### 撰寫日期：2020/04/04
### 改版日期:  
### 改版備註:

## 練習時間
假設我們資料中類別的數量並不均衡，在評估準確率時可能會有所偏頗，試著切分出 y_test 中，0 類別與 1 類別的數量是一樣的 (亦即 y_test 的類別是均衡的)

In [102]:
import numpy as np
import pandas as pd
X = np.arange(1000).reshape(200, 5)
y = np.zeros(200)
y[:40] = 1

In [103]:
print(y)
print('-----------------')
print(X.shape)
print('-----------------')
print(y.shape)
print('-----------------')
print(pd.DataFrame(y)[0].value_counts())

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0.]
-----------------
(200, 5)
-----------------
(200,)
-----------------
0.0    160
1.0     40
Name: 0, dtype: int64


可以看見 y 類別中，有 160 個 類別 0，40 個 類別 1 ，請試著使用 train_test_split 函數，切分出 y_test 中能各有 10 筆類別 0 與 10 筆類別 1 。(HINT: 參考函數中的 test_size，可針對不同類別各自作切分後再合併)

### ANS:

#### test_size
>float, int or None, optional (default=None)
If float, should be between 0.0 and 1.0 and represent the proportion of the dataset to include in the test split. If int, represents the absolute number of test samples. If None, the value is set to the complement of the train size. If train_size is also None, it will be set to 0.25.

In [104]:
from sklearn.model_selection import train_test_split
# 設定sample取樣數
sample_size = 40
# 設定測試集取樣數
test_num = 10

#取1
X_train1, X_test1, y_train1, y_test1 = train_test_split(X[:sample_size], y[:sample_size], \
                                                        test_size=test_num ,random_state=42)

#取0
X_train2, X_test2, y_train2, y_test2 = train_test_split(X[sample_size:], y[sample_size:], \
                                                        test_size=test_num ,random_state=42)

In [106]:
#顯示結果
print(pd.DataFrame(y_test1)[0].value_counts())
print('-----------------')
print(pd.DataFrame(y_test2)[0].value_counts())
print('-----------------')
y_test = np.concatenate((y_test1,y_test2))
print('合併結果')
print(y_test)
print(pd.DataFrame(y_test)[0].value_counts())

1.0    10
Name: 0, dtype: int64
-----------------
0.0    10
Name: 0, dtype: int64
-----------------
合併結果
[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
0.0    10
1.0    10
Name: 0, dtype: int64
